lib/ansible/modules/cloud/dimensiondata/dimensiondata_firewall.py

```diff
@@ -0,0 +1,510 @@
+    network_obj = get_network_domain(driver, network_domain, location)
+
+    position = position.upper()
+    # Check if relative_to_rule exists
+    if position == 'BEFORE' or position == 'AFTER':
+        if relative_to_rule is None:
+            module.fail_json(
+                msg="'relative_to_rule' must be a valid rule name when 'position' is 'BEFORE' or 'AFTER'"
+            )
+
+        target_rule = get_firewall_rule_by_name(driver, relative_to_rule,
+                                                network_obj.id)
+        if target_rule is None:
+            module.fail_json(
+                msg="Rule '%s' specified in 'relative_to_rule' not found" % relative_to_rule
+            )
+
+    # Get rule if exists
+    existing_rule = get_firewall_rule_by_name(driver, name, network_obj.id)
+
+    # Process state
+    if state == 'present' or state == 'enabled' or state == 'disabled':
+        # Is enabled?
+        enabled = 'false' if state == 'disabled' else 'true'
+        # Get SOURCE network parts
+        source_obj = to_source_or_destination(source, source_port)
+        # Get DESTINATION network parts
+        destination_obj = to_source_or_destination(destination,
+                                                   destination_port)
+        if existing_rule is None:
+            # Create Firewall Rule
+            create_firewall_rule(
+                module, driver, name, action,
+                network_obj.id,
+                ip_version, protocol,
+                source_obj.ip_address,
+                source_obj.ip_prefix_size,
+                source_obj.port_begin,
+                source_obj.port_end,
+                destination_obj.ip_address,
+                destination_obj.ip_prefix_size,
+                destination_obj.port_begin,
+                destination_obj.port_end, position,
+                relative_to_rule, enabled
+            )
+        else:
+            # Rule already exists
+            # Exit now if state doesnt need updating
+            if existing_rule.enabled == 'false' and state == 'disabled':
+                module.exit_json(changed=False,
+                                 msg="Firewall rule exists and is disabled.",
+                                 firewall_rule=rule_obj_to_dict(existing_rule))
+            elif existing_rule.enabled == 'true' and state != 'disabled':
+                module.exit_json(changed=False,
+                                 msg="Firewall rule exists and is enabled.",
+                                 firewall_rule=rule_obj_to_dict(existing_rule))
+            else:
+                # Sync Rule state
+                sync_firewall_rule_state(module, driver, existing_rule, state)
+    elif state == "absent":
+        if existing_rule is None:
+            module.exit_json(msg="Rule with name '%s' not found." % name)
+        else:
+            # Delete rule
+            delete_firewall_rule(module, driver, existing_rule)
+    else:
```

Comment:
You shouldn't need this, it's enforced by Ansible as you've already specified
`choices=['present', 'absent', 'enabled', 'disabled']),`